crates/primitives/src/fork/manager.rs

```diff
@@ -1,83 +1,66 @@
-use std::collections::VecDeque;
-
-use sov_rollup_interface::spec::SpecId;
-#[cfg(feature = "native")]
-use tracing::info;
+use sov_rollup_interface::fork::Fork;
 
 use super::ForkMigration;
 
-/// Defines the interface which a fork manager needs to implement.
-pub trait Fork {
-    /// Returns the currently active fork.
-    fn active_fork(&self) -> SpecId;
-
-    /// Register a new L2 block with fork manager
-    fn register_block(&mut self, height: u64) -> anyhow::Result<()>;
-}
-
-pub type SpecActivationBlockHeight = u64;
-
 pub struct ForkManager {
-    active_spec: SpecId,
-    specs: VecDeque<(SpecId, SpecActivationBlockHeight)>,
+    forks: Vec<Fork>,
+    active_fork_idx: usize,
     migration_handlers: Vec<Box<dyn ForkMigration + Sync + Send>>,
 }
 
 impl ForkManager {
-    pub fn new(
-        current_l2_height: u64,
-        active_spec: SpecId,
-        mut specs: Vec<(SpecId, SpecActivationBlockHeight)>,
-    ) -> Self {
-        // Filter out specs which have already been activated.
-        specs.retain(|(spec, block)| *spec != active_spec && *block > current_l2_height);
+    pub fn new(mut forks: Vec<Fork>, current_l2_height: u64) -> Self {
         // Make sure the list of specs is sorted by the block number at which they activate.
-        specs.sort_by_key(|(_, block_number)| *block_number);
+        forks.sort_by_key(|fork| fork.activation_height);
+
+        let mut active_fork_idx = 0;
+        for (idx, fork) in forks.iter().enumerate() {
+            if current_l2_height >= fork.activation_height {
+                active_fork_idx = idx;
+            } else {
+                break;
+            }
+        }
+
         Self {
-            specs: specs.into(),
-            active_spec,
+            forks,
+            active_fork_idx,
             migration_handlers: vec![],
         }
     }
 
     pub fn register_handler(&mut self, handler: Box<dyn ForkMigration + Sync + Send>) {
         self.migration_handlers.push(handler);
     }
-}
 
-impl Fork for ForkManager {
-    fn active_fork(&self) -> SpecId {
-        self.active_spec
+    pub fn active_fork(&self) -> &Fork {
+        &self.forks[self.active_fork_idx]
     }
 
-    fn register_block(&mut self, height: u64) -> anyhow::Result<()> {
-        if let Some((new_spec, activation_block_height)) = self.specs.front() {
-            if height == *activation_block_height {
-                #[cfg(feature = "native")]
-                info!("Activating fork {:?} at height: {}", *new_spec, height);
```

Comment:
This would be very useful IMO so let's restore it.

Reply:
agreed, restored.